### PR TITLE
Allow to create/update/delete formula mappings

### PR DESCRIPTION
**Pull request by mestachs**
requires api changes : https://github.com/BLSQ/orbf2/pull/221

formula-mapping

### CODE REVIEW

File: src/components/Formula/TopicBasedFormulas.js [hunk 1]
```diff
@@ -83,62 +84,181 @@ const Dhis2ElementDetails = ({ dhis2Object }) => {
...
+      setNewExternalReference(v.id);
+      setDhis2Object(dhis2LookupElement(v.id));
+    }
+  };
+
+  const handleConfirm = async () => {
+    if (cell.value.formulaMapping == undefined) {
+      // CREATE
+      try {
+        const resp = await externalApi()
```
Comment (pjaspers) [inline src/components/Formula/TopicBasedFormulas.js]: Would move this to a single method `createOrUpdate(cell_value)`, I think only the URL needs to change between the two of them.

File: src/components/Shared/DataElementComboAutocomplete.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    let active = true;
+
+    if (inputValue === "") {
+      setOptions([]);
+      return undefined;
+    }
+
+    externalApi()
+      .errorType("json")
+      .url(`/de_cocs`)
```
Comment (pjaspers) [inline src/components/Shared/DataElementComboAutocomplete.js:49]: (we could use ' instead of `)

File: src/actions/api.js [hunk 3]
```diff
@@ -25,6 +26,7 @@ export const externalApi = () => {
     "Accept-Language": i18n.language,
     Accept: "application/vnd.api+json;version=2",
     "X-Token": finalToken,
+    "X-Dhis2UserId": user.id,
```
Comment (pjaspers) [inline src/actions/api.js:29]: Nice

**Review by pjaspers — APPROVED** [review]
Haven't tried to run it locally (yet), but looks like a very welcome update.

**Comment by mestachs** [comment]
Extracted the component and reused it in the "non topic" based formulas mapping

formula-mapping-2

for wretch the method differs (put vs post), I've extracted the payload and page reload (need to find something better, but I'm not a redux guru)

I'll try to check for user role "hesabu admin" to enable or not the edition.

**Comment by mestachs** [comment]
now the screen will stay in readonly if the user doesn't have userRole "Hesabu Admin".

the role can created via
image

**Pull request merged**